### PR TITLE
refund_proposal_and_payment_aligment

### LINKED ISSUE #104: Refund functionality for Payment requests
Carrier billing API has provided APIs for performing add to bill functionality of both prepaid and postpaid users, under 1 step and 2 step payments. But most of the business cases are linked with refund related user cases as well. Currently, Carrier billing checkout API definition has not provided any direction on how to manage the refunding related operations. But considering the inter related use cases with the core functionality, can you provide guidelines/support for ,
1. Introduce Refund handling API alone with Carrier billing Camara API.
2. Introduce/enrich document with a proper way of handling Refund related use cases using same APIs with a guided parameter configurations.

Thanks,
Anushka

### LINKED ISSUE #153: Manage and Control the use of non-negative amounts
**Problem description**
Within [carrier_billing.yaml](https://github.com/camaraproject/CarrierBillingCheckOut/blob/main/code/API_definitions/carrier_billing.yaml#L918) monetary amounts MUST be positive. Negative values do not make sense and also "zero" value is not indeed a payment.

Due to this, proposed solution considers to set minimum value for "amount" fields to `0.001`. In the API Spec this would be:

`
...
        amount:
          type: number
          format: float
          **_minimum: 0.001_**
          multipleOf: 0.001
          description: Amount to be charged or reserved.
          example: 100
...
`

NOTE: 
- this would also apply in Ongoing #152
- Consideration of 400 Exception can be applied during PR discussion

**Possible evolution**
Taking advantage of Open #152, it is proposed to deal with this issue within Carrier Billing v0.3.0 and new Refund Interface.

**Alternative solution**
Include as a Fix for [carrier_billing.yaml](https://github.com/camaraproject/CarrierBillingCheckOut/blob/main/code/API_definitions/carrier_billing.yaml) v0.2.1 separately

**Additional context**

N/A

### LINKED ISSUE #154: Error Model aligment with Commonalities Meta-Release Fall24
**Problem description**
In the Context of [Commonalities/Issue 180](https://github.com/camaraproject/Commonalities/issues/180) and [Commonalities/Issue 127](https://github.com/camaraproject/Commonalities/issues/127). An error model alingment is required.

@bigludo7 also commented in 17th April meeting that could make sense to have some errors being addressed with 422 exception instead of 403 one.

Based on that aligment, it would be required to make proper adaptations in both carrier billing payment and carrier billing refund functionality.

**Possible evolution**
Pending to be proposed

**Alternative solution**
N/A

**Additional context**
N/A

### LINKED ISSUE #158: Notifications Model aligment with Commonalities Meta-Release Fall24
**Problem description**
Aligment of Carrier Billing APIs with Updated Notifications Model. Output of:
- [Commonalities/185](https://github.com/camaraproject/Commonalities/issues/185) (Also #140, #149, #153) - PRs: 189, 198
- [Commonalities/191](https://github.com/camaraproject/Commonalities/issues/191) - PR: 194

Based on that aligment, it would be required to make proper adaptations in both carrier billing payment and carrier billing refund functionality.

**Possible evolution**
Based on conclusion of above tracks

**Alternative solution**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by PedroDiez**
#### What type of PR is this?

* enhancement/feature

#### What this PR does / why we need it:

This PR covers proposal for refund functionality.

- It provides a design for both total and partial refunds, allowing for asynchronous model in business proceesing
- Events Model covered
- Exceptions Model covered
- Some alignment within commonalities included
- Considering remainingAmount not refunded in Payment resource (i.e. API)
- Considering amount (minimum:0.001) and taxAmount (minimum: 0) to be non-negative
- PhoneNumber aligment in Carrier billing. Exceptions added (draft input to check the whole model after).
- merchantIdentifier model, 422 exceptions and more description for refund
- polimorphic model for refund
- new proposal for remaining payment amount

#### Which issue(s) this PR fixes:

Initial Proposal - Baseline for:

> Fixes #104 
> Fixes #158

2024-05-15
> Fixes #153 (commit 7b224ea01003e4d8aac1be692f12c34add63c293)
> Rename status to `paymentStatus` and `refundStatus`. Also cover amount (minimum:0.001) and taxAmount (minimum: 0) to be non-negative in refunds (commit d12b4cfcf29a5358c64fcad249e9617a11500d57)

2024-05-24/25
> PhoneNumber aligment in Carrier Billing (commit bb99437f73199e8e70dd7dffd1acd838538e925a)
> merchantIdentifier model in Carrier Billing Refund, 422 Exceptions, more description (commit 659b2aa46ef4dd0494a8576e47952cea64761601)
> polymorphic refund model (commit 08837a86b7ab472aa9d84c7bbcab8cdff022ca5b)
> new proposal for remaining payment amount (commit efe24441d606325a7b55d1c24a464f1a54b4a209)

2024-06-10/11/12
> Alignment with Commonalities PR#214 (x-camara-commonalities header) and Notifications CloudEvent Model (commit 4d03b249acb8eddfb15a12681897ff5c936eb0e3)
> Inclusion of `paymentItemId` concept ( commit 9c723cd6c8f37cbf5ff6dac146d2343b6dffc5e4)
> fixing some typos and missings in error exceptions (commits e2be5764777e87a7ab10771617424fcd3b7db7a9, 8d079e383cdeec649e63cdc07b2c646099a81ae1 and a647deaa050492aae4a662710decddbb0a764023)

2024-06-19
> Description enhancements and description.info alignment with Commonalities ICM output (# Authorization and authentication section) (commit 85a008a0a6c619a6521dd2a759ae4c161fbffcfc)
> Semantic enhancements (denied refund in ASYNC context, remainingAmount behaviour, based on Ludovic comments/suggestions) (commit 9377d6cacf22c5e182ce9508f89c80bc8d777f34)

2024-06-21
Fixes #154 
> Commonalities alignment errors (#154) and endpoint versions (commit 8c230ba20700a15605abd94f748fcebc6ea658f2)

2024-06-27
> Fix typo un refund-completed event (paymentDate should read refundDate) (commit 8cfe775ce8e556e15094eb8fe0d92c9e1932d7f4)

### CODE REVIEW

**Review by bigludo7 — COMMENTED** [review]
Hello @PedroDiez 

I begun the review and few comments to share with you. Most of this points are not r_equest for change_ but question I got when I did this review.

- Are we sure to manage `refund_in_progress`, `totally_refunded`, `partially_refunded` as TransactionOperationStatus ? my point is that a payment could have been Succeeded and then refunded. Both state are valid. Need to be sure we did not need to tackle that in a specific `refundStatus`.
- I'm still not sure if refund should be managed in a separate yaml or same yaml that payment. Have you @PedroDiez good rationale to make them separate?
- Are we sure that refundAmount should be mandatory ? if `type` is `total` this could be source of unnecessary error (what about a request with `type:total` and amount inconsistent btw the payment and the refund?).
- If requester fill refundDetails do we need to make check vs payment item? I guess yes.
- In payment we require a `referenceCode` in the request - do we need one here for the refund?

**Comment by bigludo7** [comment]
One quick question.... what is the correlation between refundDetails.refundItem and paymentDetail.paymentItem ? I guess we expect a strong correlation meaning that if I have 3 payment item I could have a refund only for the second item right?
As such in order to have this correlation I'm wondering if we should not add a paymentItem.id in paymentItem and refer this in refundItem.
Perhaps this is 'too much' but got this comment in looking in refundItem and how to use it.

**Review by bigludo7 — COMMENTED** [review]
Hi @PedroDiez 
We're almost there. We have changed the authorization sentences in ICM for the documentation. 
also we have probably to have a rule for the remaining amount regarding processing refund.
nothing big here.

File: code/API_definitions/carrier_billing.yaml [hunk 1]
```diff
@@ -71,36 +71,42 @@ info:
...
 
-    After `confirmPayment`, status=`succeeded` OR `denied`, depending whether it was successful or not.
+    After `confirmPayment`, paymentStatus=`succeeded` OR `denied`, depending whether it was successful or not.
 
-    After `cancelPayment`, status=`cancelled`.
+    After `cancelPayment`, paymentStatus=`cancelled`.
+
+    # Authorization and authentication
+
+    CAMARA guidelines defines a set of authorization flows which can grant API clients access to the API functionality, as outlined in the document [CAMARA-API-access-and-user-consent.md](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-API-access-and-user-consent.md). Which specific authorization flows are to be used will be determined during onboarding process, happening between the API Client and the Telco Operator exposing the API, taking into account the declared purpose for accessing the API, while also being subject to the prevailing legal framework dictated by local legislation.
```
[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing.yaml:102]: let me check Ludovic, as also I think commonalities have been updated

File: code/API_definitions/carrier_billing_refund.yaml [hunk 2]
```diff
@@ -0,0 +1,1308 @@
+openapi: 3.0.3
+info:
+  description: |-
+    Service Enabling Refunds against OB Carrier Billing Systems
```
[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing_refund.yaml:4]: DONE

File: code/API_definitions/carrier_billing.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing.yaml:102]: DONE

File: code/API_definitions/carrier_billing_refund.yaml [hunk 3]
```diff
@@ -0,0 +1,1308 @@
...
+
+    If `createRefund` is an **ASYNC** process:
+    - Response contains `refundId` and refundStatus=`processing`. After completion:
+      - When refund is successfully completed then refundStatus=`succeeded`.
+      - When refund is not successfully performed then refundStatus=`denied`.
+    - In case of any error scenario `refundId` is not created.
+
+    # Authorization and authentication
+
+    CAMARA guidelines defines a set of authorization flows which can grant API clients access to the API functionality, as outlined in the document [CAMARA-API-access-and-user-consent.md](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-API-access-and-user-consent.md). Which specific authorization flows are to be used will be determined during onboarding process, happening between the API Client and the Telco Operator exposing the API, taking into account the declared purpose for accessing the API, while also being subject to the prevailing legal framework dictated by local legislation.
```
[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing_refund.yaml:46]: DONE

File: code/API_definitions/carrier_billing_refund.yaml [hunk 4]
```diff
@@ -0,0 +1,1308 @@
...
+
+    # API Functionality
+
+    This API allows to third party clients to request the refund of a given payment, as well as to retrieve information about a specific refund or a list of refunds.
+
+    The API provides several endpoints/operations:
+    - An endpoint to request a refund, named `createRefund`.
+    - A set of endpoints to retrieve information about a list of refunds or a specific refund (identified by its specific `refundId`), named `retrieveRefunds` and `retrieveRefund` respectively.
+    - A callback endpoint where API Server can send notifications about a refund procedure, as defined within `createRefund` operation, towards the `sink` when provided by API client.
+
```
[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing_refund.yaml:30]: DONE

File: code/API_definitions/carrier_billing_refund.yaml [hunk 5]
```diff
@@ -0,0 +1,1308 @@
...
+    get:
+      security:
+        - openId:
+            - carrier-billing-refund:refunds:read
+      tags:
+        - Refund
+      summary: Get remaining amount not refunded for a given payment
+      operationId: retrievePaymentRemainingAmount
+      description: |-
+        Retrieve remaining amount not yet refunded for a given payment.
```
[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing_refund.yaml:299]: Get your point, yes probably a note indicating that is needed to clarify and align the implementation (a refund in progress is an amount not possible to be request for a refund until a final consolidation is performed)

File: code/API_definitions/carrier_billing_refund.yaml [hunk 6]
```diff
@@ -0,0 +1,1308 @@
...
+
+    If `createRefund` is a **SYNC** process:
+    - Response contains `refundId` and refundStatus=`succeeded`.
+    - In case of any error scenario `refundId` is not created.
+
+    If `createRefund` is an **ASYNC** process:
+    - Response contains `refundId` and refundStatus=`processing`. After completion:
+      - When refund is successfully completed then refundStatus=`succeeded`.
+      - When refund is not successfully performed then refundStatus=`denied`.
+    - In case of any error scenario `refundId` is not created.
```
[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing_refund.yaml:42]: Yes, have the same understanding, proposing some additional description

[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing_refund.yaml]: DONE

File: code/API_definitions/carrier_billing_refund.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing_refund.yaml]: DONE

**Review by bigludo7 — APPROVED** [review]
LGTM/

**Review by bigludo7 — APPROVED** [review]
LGTM/

**Comment by PedroDiez** [comment]
Merging it as commenting in Meeting minutes. Initial wip proposal

**Pull request merged**